### ISSUE TITLE
YouTube: "Upscale" behaves the same as "Stretch" and is stretching the frame, rather than cropping in.

### ISSUE THREAD

**Issue by WinkelCode** [dvlden/ultrawideo#94]
### Describe the bug
"Upscale" mode doesn't work on YouTube anymore.

### Reproduction steps
1. Go to https://www.youtube.com/watch?v=Qono7fLH9C8
2. Set mode to "Upscale"
3. Go full screen

### Expected behavior
Video is cropped to fill the screen

### Actual behavior
Video is stretched, identical to "Stretch" mode

### Environment
- **OS**: Windows 11 22H2
- **Browser**: Microsoft Edge: `Version 109.0.1518.55 (Official build) (64-bit)`

**Edit:** It works as expected on Netflix

**Edit 2:** When cycling with a hotkey, there is no perceivable difference between "Upscale" and "Stretch", both are stretched in the same manner.

**Comment by dvlden** [comment]
Can't reproduce at the moment. I am on macOS (latest) + Chrome (latest).
Will check back soon with BrowserStack.

**Comment by MultiChipModule** [comment]
This is the same for disney+ on chrome

**Comment by dvlden** [comment]
> This is the same for disney+ on chrome

Can't reproduce. Working properly on macOS @ Chrome.

**Comment by dvlden** [comment]
Unable to do it via BrowserStack. It's too laggy for desktop tests.

**Comment by markey** [comment]
I can confirm it stopped working with Edge on W11.

**Comment by Axonen** [comment]
I have the same issue for about a week. Upscale doesn't work and it looks exactly like "Stretch". 
Chrome, Windows 10

**Comment by viktorcheb** [comment]
Same issue YouTube, video is stretched instead of upscaled on Chrome Windows 11.

**Comment by selfmadebuba** [comment]
Same thing here Win10 & Win11
Updated chrome on both PCs
Ultrawide is streaching only but upper and lower black bars stay.
problem exist on:
Disney+
Youtube
Netflix

**Comment by dvlden** [comment]
Hello everyone.

I encourage you to go ahead and check this guide and post again or edit existing comment of yours, if you make it work.

https://github.com/dvlden/ultrawideo/discussions/62

**Comment by WinkelCode** [comment]
I discovered some stuff:

1. It seems to sometimes crop in correctly when initially opening the video.
2. It actually works when the dev tools are open in the same window as the video (i.e. not as a separate window)

**Comment by selfmadebuba** [comment]
> I discovered some stuff:
> 
> 1. It seems to sometimes crop in correctly when initially opening the video.
> 2. It actually consistently works when the Dev Tools are open in the same window as the video (i.e. not as a separate window). It stops working when closing, or "popping out" the Dev Tools, resizing the screen area for the tools crops the video in and out without issues, though having the dev tools open is distracting of course.
> 
> @dvlden I wonder if a solution would be to somehow emulate this? I doubt it's specifically a Dev Tools thing, probably more related to sharing the window with something else?

I have the same effect when I turn off ultravideo plugin in chrome. Check if you have same thing

**Comment by WinkelCode** [comment]
> I have the same effect when I turn off ultravideo plugin in chrome. Check if you have same thing

What are you doing exactly?

**Comment by selfmadebuba** [comment]
When you turn off ultravideo plugin in chrome extensions. Restart. 
Use i.e. Netflix. full screen will stretch video on my 21:9 in full screen. 
But when I turn devtools in fullscreen video is correct - not stretched.

**Comment by WinkelCode** [comment]
> When you turn off ultravideo plugin in chrome extensions. Restart. Use i.e. Netflix. full screen will stretch video on my 21:9 in full screen. But when I turn devtools in fullscreen video is correct - not stretched.
> 
> EDIT: I watched The Ninth Gate movie. It can be important because Netflix can optimise some movies and some not. So we should test on the same platform same movie basis.

I don't fully understand your steps, is UltraWideo enabled or disabled when you are on Netflix?

I've been using 4K 16:9 YouTube videos for testing, does it behave the same way for you?

**Comment by selfmadebuba** [comment]
> > When you turn off ultravideo plugin in chrome extensions. Restart. 

I think it's clear in the first line.

> I've been using 4K 16:9 YouTube videos for testing, does it behave the same way for you?
I have ultaWide monitor 21:9. I do not recollect problems with standard 16:9 videos. Give me example video and it's behavior so i test it on my 21:9

**Comment by WinkelCode** [comment]
> I have ultaWide monitor 21:9. I do not recollect problems with standard 16:9 videos. I have problems with 21:9 videos. Give me example video and it's behavior so i test it on my 21:9

The video from my original post: https://www.youtube.com/watch?v=Qono7fLH9C8

It's important to note that the type of 21:9 videos you'd need this plugin for are actually also 16:9, they need to be cropped in because the black bars are "baked into" the video.

**Comment by selfmadebuba** [comment]
![TEST02](https://user-images.githubusercontent.com/10959671/219459093-04a71d6e-466c-460e-91e8-f0a706a0d9e5.jpg)

**Comment by WinkelCode** [comment]
> ![TEST02](https://user-images.githubusercontent.com/10959671/219459093-04a71d6e-466c-460e-91e8-f0a706a0d9e5.jpg)
> 
> without ultravideo utravideo extiension off chrome without any changes to flags win11 nvidia mobile gpu

The goal is to make the video fill the screen by cropping the top and bottom.

**Comment by selfmadebuba** [comment]
> So you are just simply not using UltraWide? If it's disabled*, it won't do anything, just as if it's not installed to begin with.
> 
> **Edit 2:** *If it's disabled and then the page is refreshed, and especially if the browser is restarted.

Of course it wont do anything. I'm making a point that whatever you do use ultravideo or not. DEVTOOLS do the same.

**Comment by WinkelCode** [comment]
> > So you are just simply not using UltraWide? If it's disabled*, it won't do anything, just as if it's not installed to begin with.
> > **Edit 2:** *If it's disabled and then the page is refreshed, and especially if the browser is restarted.
> 
> Of course it wont do anything. I'm making a point that whatever you do use ultravideo or not. DEVTOOLS do the same.

I am testing on Microsoft Edge (`110.0.1587.46`) running on Windows 11 (Edge is pretty much a reskin of Chrome anyway), Dev Tools only fix the stretching if I also have UltraWideo enabled in the "upscale" mode. Pretty much having Dev Tools open in the same window as the fullscreen video "magically fixes" the problem and it all works as expected. Dev Tools don't do anything to the video.

**Comment by selfmadebuba** [comment]
> The goal is to make the video fill the screen by cropping the top and bottom.

My goal:  Do not strech 21:9 video on 21:9 I cant give screenshot because Netflix video is impossible to screenshot. It makes 21:9 video on 21:9 screeen as streched with two black bars: Up and down. It should fill the screen.
![TEST03](https://user-images.githubusercontent.com/10959671/219461342-ed2a138c-f28b-4aa4-8e47-8a8c22192b31.png)

**Comment by WinkelCode** [comment]
> > The goal is to make the video fill the screen by cropping the top and bottom.
> 
> My goal: Do not strech 21:9 video on 21:9 I cant give screenshot because Netflix video is impossible to screenshot. It makes 21:9 video on 21:9 screeen as streched with two black bars: Up and down. It should fill the screen. ![TEST03](https://user-images.githubusercontent.com/10959671/219461342-ed2a138c-f28b-4aa4-8e47-8a8c22192b31.png)

Try this video: https://www.youtube.com/watch?v=BmllggGO4pM

Does it behave like Netflix? It has baked in black bars.

**Comment by selfmadebuba** [comment]
ULTRAVIDEO ON. Cheome flags default.

![TEST04](https://user-images.githubusercontent.com/10959671/219463258-d79d16f6-9983-4419-8211-77bb43f19874.png)

**Comment by WinkelCode** [comment]
@selfmadebuba Yeah that's the bug: "Upscale" is the same as "Stretch". 

Have you found a way to make "Upscale" properly crop out the top and bottom black bars?

As I mentioned before, the Dev Tools, at least on Microsoft Edge, somehow make it so UltraWideo is suddenly able to crop the video again just by having it open in the same window as (to the right or below) the fullscreen video.

**Comment by selfmadebuba** [comment]
If you use Approach 2 and 7change graphics backend in flags for ANGLE to DX9 or OPENGL it starts working.

I don't like this solution. It only works that way because ultravideo doesn't support dx11+ which is a shame.

Changing to DX9/OPENGL worsens the performance of the entire web browser in terms of speed and battery consumption.

That's a shitty and lazy solution and not for a long run.

**Comment by dvlden** [comment]
I like that you two are trying to solve this issue. 
You can read more about this issue by an alternate extension author:

[Ultrawidify](https://stuff.tamius.net/sacred-texts/2021/08/19/ultrawidify-and-chrome-2021-edition-episode-2/)

**TLDR:**
Google messed up the hardware acceleration for DirectX11 and DirectX12, which are different ways for programs to tell a graphics card what to do. Hardware acceleration using different versions of DirectX can result in different bugs, and this issue affects various browsers that copy Google's approach, except for Firefox. To display high-resolution videos, graphics cards can decode video faster than processors by using their many simple cores, but programs need to use the right API to tell the graphics card what to do.

---

So, as this has nothing to do with the extension itself, as extension does not rely on DX, but browser does on some things, this is not something that is fixable. Changing ANGLE to DX9 is one of the solutions that simply works, until this gets fixed at some point by the whoever is responsible for it.

**Comment by selfmadebuba** [comment]
I do not agree.
It's hard to name it as google chrome bug... It's different implementation and it has different effect. Issue has been archived 7 months ago by google and won't be worked or acknowledged as a bug in the future. I's how new API transform function implementation works. That's it. Not a bug. A change.

Another thing you said that there are many different API's around. Well there is only one importand one. The default one which will be for a future in development. Default one which works as default in all chromium implementations from today to the future.

I don't thing is unfixable. It's unfixable using old method based on simple transform.
But eventually there is a different video cropping and transforming possible.
So technically it's fixable but by different method. 
Changing to DX9 is a road to nowhere. Users can't get stuck on old dx9 on their browser. DX that hasn't been developed for years and lose performance and functionality until someone disables old dx9 and make an end to the story.

**Comment by dvlden** [comment]
This extension works with simple CSS rules that are pretty much future-proof. 

Unlike some other extensions, it does not use JS to calculate width and height and track resize events to re-calculate which can be pretty heavy. That method probably works for DirectX problems, but that is not how this extension works internally and that will not change. 

I do not know how DevTools would affect this anyhow, so that's not really helpful for me.
If Windows, DirectX or Chromium have issues with `object-fit` property of CSS, that's not on me to fix.

**Comment by WinkelCode** [comment]
To me it looks like the Dev Tools somehow force the video to use play in a way that is accessible to UltraWideo, given how popular both YouTube and Chromium-based browsers are, I think it would be worthwhile to look into whether it is possible to replicate this in some other way. It doesn't see too far fetched to me that it could be possible to achieve this without having a visible element in screen. I wonder if it might be possible to create an overlay that causes the same behavior as the Dev Tools, but something like transparent and 1x1 pixels in size.

**Comment by selfmadebuba** [comment]
How the @% it's completed now? I didn't notice any solution to problem...

It doesn't work and nothing has been done. So let's mark it as COMPLETED

Good job...

**Comment by dvlden** [comment]
It's rather a browser, graphics or OS issue. Extension can't do anything about it. 
If the mode behaves different than intended by the CSS Spec., there's nothing that can be done from my end.

Aside, I can't reproduce it on macOS / Chrome, nor macOS / Firefox.

Since it's an old issue and nobody found the solution, I think it's best to leave it closed.
